File: HEAPS/minheap.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
void swap(int* x, int* y) {
    int temp = *x;
    *x = *y;
    *y = temp;
}
struct MinHeap {
    int* heap;
    int capacity;
    int heap_size;
};
struct MinHeap* createMinHeap(int capacity) {
    struct MinHeap* minHeap = (struct MinHeap*)malloc(sizeof(struct MinHeap));
    minHeap->heap = (int*)malloc(capacity * sizeof(int));
    minHeap->capacity = capacity;
    minHeap->heap_size = 0;
    return minHeap;
}
int parent(int i) {
    return (i - 1) / 2;
}
int left(int i) {
    return 2 * i + 1;
}
int right(int i) {
    return 2 * i + 2;
}
void minHeapify(struct MinHeap* minHeap, int i) {
    int smallest = i;
    int leftChild = left(i);
    int rightChild = right(i);
    if (leftChild < minHeap->heap_size && minHeap->heap[leftChild] < minHeap->heap[smallest])      
      smallest = leftChild;
    if (rightChild < minHeap->heap_size && minHeap->heap[rightChild] < minHeap->heap[smallest])  
      smallest = rightChild;
    if (smallest != i) {
        swap(&minHeap->heap[i], &minHeap->heap[smallest]);
        minHeapify(minHeap, smallest);
    }
}

int extractMin(struct MinHeap* minHeap) {
    if (minHeap->heap_size <= 0)
        return INT_MAX;
    if (minHeap->heap_size == 1) {
        minHeap->heap_size--;
        return minHeap->heap[0];
    }
    int root = minHeap->heap[0];
    minHeap->heap[0] = minHeap->heap[minHeap->heap_size - 1];
    minHeap->heap_size--;
    minHeapify(minHeap, 0);
    return root;
}
void decreaseKey(struct MinHeap* minHeap, int i, int new_val) {
    minHeap->heap[i] = new_val;
    while (i != 0 && minHeap->heap[parent(i)] > minHeap->heap[i]) {
        swap(&minHeap->heap[i], &minHeap->heap[parent(i)]);
        i = parent(i);
    }
}
void deleteKey(struct MinHeap* minHeap, int i) {
    decreaseKey(minHeap, i, INT_MIN);
    extractMin(minHeap);
}
void insertKey(struct MinHeap* minHeap, int k) {
    if (minHeap->heap_size == minHeap->capacity) {
        printf("\nOverflow: Could not insertKey\n");
        return;
    }
    int i = minHeap->heap_size;
    minHeap->heap_size++;
    minHeap->heap[i] = k;
    while (i != 0 && minHeap->heap[parent(i)] > minHeap->heap[i]) {
        swap(&minHeap->heap[i], &minHeap->heap[parent(i)]);
        i = parent(i);
    }
}
int getMin(struct MinHeap* minHeap) {
    if (minHeap->heap_size > 0)
        return minHeap->heap[0];
    else
        return INT_MAX;
}
int main() {
    struct MinHeap* minHeap = createMinHeap(11);
    insertKey(minHeap, 3);
    insertKey(minHeap, 2);
    deleteKey(minHeap, 1);
    insertKey(minHeap, 15);
    insertKey(minHeap, 5);
    insertKey(minHeap, 4);
    insertKey(minHeap, 45);
    
    printf("%d ", extractMin(minHeap));
    printf("%d ", getMin(minHeap));
    
    decreaseKey(minHeap, 2, 1);
    printf("%d", getMin(minHeap));
    
    return 0;
}

File: HEAPS/minheap.py
def find_min(heap):
    return heap[0]
def create_heap():
    return []
def insert(x, heap):
    heap.append(x)
    bubble_up(len(heap) - 1, heap)
def extract_min(heap):
    heap[0], heap[-1] = heap[-1], heap[0]
    min_val = heap.pop()
    bubble_down(0, heap)
    return min_val
def decrease_key(p, delta, heap):
    heap[p] -= delta
    bubble_up(p, heap)
def merge(heap1, heap2):
    merged_heap = heap1 + heap2
    heapify(merged_heap)
    return merged_heap
def bubble_up(index, heap):
    parent = (index - 1) // 2
    while index > 0 and heap[index] < heap[parent]:
        heap[index], heap[parent] = heap[parent], heap[index]
        index = parent
        parent = (index - 1) // 2
def bubble_down(index, heap):
    left_child = 2 * index + 1
    right_child = 2 * index + 2
    smallest = index
    if left_child < len(heap) and heap[left_child] < heap[smallest]:
        smallest = left_child
    if right_child < len(heap) and heap[right_child] < heap[smallest]:
        smallest = right_child
    if smallest != index:
        heap[index], heap[smallest] = heap[smallest], heap[index]
        bubble_down(smallest, heap)
def heapify(heap):
    n = len(heap)
    for i in range((n // 2) - 1, -1, -1):
        bubble_down(i, heap)